File: spike/correlation-fastapi/src/codemap.py
"""
Controller -> service walk for FastAPI.

THE DIVERGENCE FROM NESTJS
--------------------------
Nest injects through a constructor:

    constructor(private readonly invoiceService: InvoiceService) {}
    -> resolve from the CONSTRUCTOR PARAMETER's declared type

FastAPI injects through the handler signature:

    def find_one(..., service: InvoiceService = Depends(get_invoice_service))
    -> resolve from the HANDLER PARAMETER's annotation

Different syntactic location, identical strategy: the DECLARED TYPE is the
resolution key, and it is available without a type checker or an import
graph. Parity holds.
"""

import ast
import os
from dataclasses import dataclass, field
from typing import Optional

MAX_DEPTH = 4


@dataclass
class SourceLocation:
    file: str
    enclosing_symbol: str        # 'InvoiceService.find' — the stable key
    line: int


@dataclass
class CallEdge:
    from_symbol: str
    to_symbol: str
    via: str                     # the injected parameter, e.g. 'service'
    location: SourceLocation


@dataclass
class CodeMapEntry:
    handler_symbol: str
    handler_location: SourceLocation
    chain: list[CallEdge] = field(default_factory=list)
    reachable_symbols: list[str] = field(default_factory=list)


class CodeMap:
    def __init__(self, root: str):
        self.root = os.path.abspath(root)
        self._modules: dict[str, ast.Module] = {}
        self._files: dict[str, str] = {}
        self._load()

    # ---------------------------------------------------------------- load
    def _load(self) -> None:
        for dirpath, dirnames, filenames in os.walk(self.root):
            dirnames[:] = [d for d in dirnames if d not in {"__pycache__", ".venv", "node_modules"}]
            for fn in filenames:
                if not fn.endswith(".py"):
                    continue
                full = os.path.join(dirpath, fn)
                rel = os.path.relpath(full, self.root).replace(os.sep, "/")
                try:
                    src = open(full, encoding="utf-8").read()
                    self._modules[rel] = ast.parse(src)
                    self._files[rel] = src
                except (SyntaxError, UnicodeDecodeError):
                    continue

    # ------------------------------------------------------------- lookups
    def _find_class(self, name: str) -> Optional[tuple[str, ast.ClassDef]]:
        for rel, tree in self._modules.items():
            for node in tree.body:
                if isinstance(node, ast.ClassDef) and node.name == name:
                    return rel, node
        return None

    def _find_method(self, cls: ast.ClassDef, name: str) -> Optional[ast.FunctionDef]:
        for node in cls.body:
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) and node.name == name:
                return node
        return None

    def _find_function(self, module_rel: str, name: str) -> Optional[ast.FunctionDef]:
        tree = self._modules.get(module_rel)
        if not tree:
            return None
        for node in tree.body:
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) and node.name == name:
                return node
        return None

    # ------------------------------------------------------------ analysis
    @staticmethod
    def _annotated_params(fn: ast.FunctionDef) -> dict[str, str]:
        """
        Maps parameter name -> annotated class name.

        Covers both FastAPI idioms:
            service: InvoiceService = Depends(get_invoice_service)
            service: Annotated[InvoiceService, Depends(get_invoice_service)]
        """
        out: dict[str, str] = {}
        args = fn.args
        for arg in [*args.posonlyargs, *args.args, *args.kwonlyargs]:
            ann = arg.annotation
            if ann is None:
                continue
            name = CodeMap._type_name(ann)
            if name:
                out[arg.arg] = name
        return out

    @staticmethod
    def _type_name(ann: ast.expr) -> Optional[str]:
        if isinstance(ann, ast.Name):
            return ann.id
        # Annotated[InvoiceService, Depends(...)] -> take the first arg
        if isinstance(ann, ast.Subscript):
            base = ann.value
            if isinstance(base, ast.Name) and base.id == "Annotated":
                sl = ann.slice
                inner = sl.elts[0] if isinstance(sl, ast.Tuple) and sl.elts else sl
                return CodeMap._type_name(inner)
        if isinstance(ann, ast.Attribute):
            return ann.attr
        return None

    def _method_location(self, rel: str, class_name: str, method: ast.FunctionDef) -> SourceLocation:
        return SourceLocation(
            file=rel,
            enclosing_symbol=f"{class_name}.{method.name}",
            line=method.lineno,
        )

    def build_for_handler(self, module_rel: str, function_name: str) -> Optional[CodeMapEntry]:
        handler = self._find_function(module_rel, function_name)
        if handler is None:
            return None

        handler_symbol = f"{module_rel.removesuffix('.py').replace('/', '.')}.{function_name}"
        handler_location = SourceLocation(
            file=module_rel, enclosing_symbol=handler_symbol, line=handler.lineno
        )

        chain: list[CallEdge] = []
        reachable = {handler_symbol}

        def walk(fn: ast.FunctionDef, from_symbol: str, params: dict[str, str], depth: int) -> None:
            if depth > MAX_DEPTH:
                return

            for node in ast.walk(fn):
                if not isinstance(node, ast.Call):
                    continue
                func = node.func
                if not isinstance(func, ast.Attribute):
                    continue

                receiver = func.value
                # `service.find(...)` where `service` is an annotated param,
                # or `self.x.find(...)` inside a service method.
                if isinstance(receiver, ast.Name):
                    param_name = receiver.id
                elif (
                    isinstance(receiver, ast.Attribute)
                    and isinstance(receiver.value, ast.Name)
                    and receiver.value.id == "self"
                ):
                    param_name = receiver.attr
                else:
                    continue

                target_class = params.get(param_name)
                if not target_class:
                    continue

                found = self._find_class(target_class)
                if not found:
                    continue
                target_rel, target_cls = found

                target_method = self._find_method(target_cls, func.attr)
                if target_method is None:
                    continue

                location = self._method_location(target_rel, target_class, target_method)
                to_symbol = location.enclosing_symbol
                if to_symbol in reachable:
                    continue

                reachable.add(to_symbol)
                chain.append(
                    CallEdge(
                        from_symbol=from_symbol,
                        to_symbol=to_symbol,
                        via=param_name,
                        location=location,
                    )
                )

                walk(target_method, to_symbol, self._annotated_params(target_method), depth + 1)

        walk(handler, handler_symbol, self._annotated_params(handler), 0)

        return CodeMapEntry(
            handler_symbol=handler_symbol,
            handler_location=handler_location,
            chain=chain,
            reachable_symbols=sorted(reachable),
        )

    def build_for_routes(self, routes) -> dict[str, CodeMapEntry]:
        out: dict[str, CodeMapEntry] = {}
        for route in routes:
            module_rel = route.handler_module.replace(".", "/") + ".py"
            entry = self.build_for_handler(module_rel, route.handler_function)
            if entry:
                out[route.handler_symbol] = entry
        return out

File: spike/correlation/src/codemap.ts
import { Project, SyntaxKind, ClassDeclaration, MethodDeclaration, SourceFile } from 'ts-morph';
import * as path from 'path';
import { CodeMapEntry, CallEdge, SourceLocation, RouteMapping } from './types';

const MAX_DEPTH = 4;

export class CodeMap {
  private project: Project;
  private root: string;

  constructor(tsConfigFilePath: string) {
    this.project = new Project({ tsConfigFilePath });
    this.root = path.dirname(tsConfigFilePath);
  }

  private rel(file: SourceFile): string {
    return path.relative(this.root, file.getFilePath()).split(path.sep).join('/');
  }

  private findClass(name: string): ClassDeclaration | undefined {
    for (const sf of this.project.getSourceFiles()) {
      const cls = sf.getClass(name);
      if (cls) return cls;
    }
    return undefined;
  }

  /**
   * Maps constructor-injected property names to their declared class.
   *
   *   constructor(private readonly invoiceService: InvoiceService) {}
   *     -> { invoiceService: 'InvoiceService' }
   *
   * This is what lets `this.invoiceService.find()` resolve to a real symbol.
   * NestJS DI is constructor-based by convention, so the declared parameter
   * type is a reliable, purely syntactic resolution — no type checker needed,
   * which keeps this fast and robust on partial codebases.
   */
  private injectedProperties(cls: ClassDeclaration): Map<string, string> {
    const map = new Map<string, string>();
    const ctor = cls.getConstructors()[0];
    if (!ctor) return map;

    for (const param of ctor.getParameters()) {
      const typeNode = param.getTypeNode();
      if (!typeNode) continue;
      map.set(param.getName(), typeNode.getText());
    }
    return map;
  }

  private locationOf(method: MethodDeclaration, className: string): SourceLocation {
    return {
      file: this.rel(method.getSourceFile()),
      enclosingSymbol: `${className}.${method.getName()}`,
      line: method.getStartLineNumber(),
    };
  }

  /**
   * Walks outward from a route handler through injected-service calls.
   *
   * The controller is almost never where an authorization bug lives, so a
   * correlation that stops at the handler points reviewers at the wrong file.
   * This is the hop that makes the finding actionable.
   */
  buildForHandler(controllerClass: string, handlerMethod: string): CodeMapEntry | null {
    const cls = this.findClass(controllerClass);
    if (!cls) return null;

    const handler = cls.getMethod(handlerMethod);
    if (!handler) return null;

    const handlerLocation = this.locationOf(handler, controllerClass);
    const chain: CallEdge[] = [];
    const reachable = new Set<string>([handlerLocation.enclosingSymbol]);

    const walk = (
      currentClass: ClassDeclaration,
      currentClassName: string,
      currentMethod: MethodDeclaration,
      depth: number,
    ): void => {
      if (depth > MAX_DEPTH) return;

      const injected = this.injectedProperties(currentClass);
      const fromSymbol = `${currentClassName}.${currentMethod.getName()}`;

      for (const call of currentMethod.getDescendantsOfKind(SyntaxKind.CallExpression)) {
        const callee = call.getExpression();
        if (!callee.isKind(SyntaxKind.PropertyAccessExpression)) continue;

        const calledMethodName = callee.getName();          // 'find'
        const receiver = callee.getExpression();            // 'this.invoiceService'
        if (!receiver.isKind(SyntaxKind.PropertyAccessExpression)) continue;
        if (receiver.getExpression().getKind() !== SyntaxKind.ThisKeyword) continue;

        const propertyName = receiver.getName();            // 'invoiceService'
        const targetClassName = injected.get(propertyName);
        if (!targetClassName) continue;

        const targetClass = this.findClass(targetClassName);
        if (!targetClass) continue;

        const targetMethod = targetClass.getMethod(calledMethodName);
        if (!targetMethod) continue;

        const location = this.locationOf(targetMethod, targetClassName);
        const toSymbol = location.enclosingSymbol;

        if (reachable.has(toSymbol)) continue;              // cycle guard
        reachable.add(toSymbol);
        chain.push({ fromSymbol, toSymbol, via: propertyName, location });

        walk(targetClass, targetClassName, targetMethod, depth + 1);
      }
    };

    walk(cls, controllerClass, handler, 0);

    return {
      handlerSymbol: handlerLocation.enclosingSymbol,
      handlerLocation,
      chain,
      reachableSymbols: [...reachable],
    };
  }

  buildForRoutes(routes: RouteMapping[]): Map<string, CodeMapEntry> {
    const out = new Map<string, CodeMapEntry>();
    for (const route of routes) {
      const entry = this.buildForHandler(route.controllerClass, route.handlerMethod);
      if (entry) out.set(route.handlerSymbol, entry);
    }
    return out;
  }
}
